### PR TITLE
Ensemble attack: Meta classifier pipeline 

### PR DESCRIPTION
**Pull request by sarakodeiri**
# Short Description

Clickup Ticket(s): (https://app.clickup.com/t/868fm2hg6)

Both choices of meta classifiers can be trained and saved as pickle files. Predictions can also be done and evaluated with using TPR@FPR. The PR includes preprocessing methods for getting training features as well (e.g. Gower distance, the DOMIAS method.) 

Things to note:
* The pipeline has been tested with the data provided by the original repository which is very small in size. Other thorough tests are required when all parts of the pipeline are finalized.
* The original repository has an "iteration" count and doesn't use it properly. The same has been implemented here.
* This implementation is loyal to the original repository about DOMIAS as well, but it might be better if DOMIAS is used only for non-categorical data, which is something to be tested after the pipeline has been finalized as well. 

# Tests Added

Tests have been added for appropriate model fitting, data preprocessing, and the prediction function.

### CODE REVIEW

File: examples/ensemble_attack_example/real_data_collection.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+from enum import Enum
+from pathlib import Path
+
+import pandas as pd
+from omegaconf import DictConfig
+
+from midst_toolkit.attacks.ensemble.data_utils import load_dataframe, save_dataframe
+
+
+class AttackType(Enum):
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply examples/ensemble_attack_example/real_data_collection.py:15]: @fatemetkl

File: examples/ensemble_attack_example/real_data_collection.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        "final",
+    ], "data_split should be one of 'train', 'dev', or 'final'."
+    # `data_id` is the folder numbering of each training or challenge dataset,
+    #  and is defined with the provided config.
+    data_id = expand_ranges(data_processing_config.folder_ranges[data_split])
+
+    # Get file name based on the kind of dataset to be collected (i.e. train vs challenge).
+    # TODO: Make the below parsing a bit more robust and less brittle
+    generation_name = attack_type.value.split("_")[0]
+    if dataset == "challenge":
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply examples/ensemble_attack_example/real_data_collection.py:75]: @fatemetkl

File: examples/ensemble_attack_example/config.yaml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+              "8",
+              "9",
+              "10",
+              "11",
+              "12",
+              "13",
+          ]
+
+
+# Training settings (placeholder)
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply examples/ensemble_attack_example/config.yaml:94]: For the "Training" settings, I'm assuming the numbers we've assigned are placeholder numbers. But for the "Metaclassifier" settings, I'm not sure if I want to keep the "epoch" variable because it's not used in the current implementation. Couldn't think of a better name that would capture the temporary nature of it.

File: src/midst_toolkit/attacks/ensemble/blending.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        # 2. Get DOMIAS predictions
+        domias_features = calculate_domias(df_input=df_input, df_synth=df_synth, df_ref=df_ref)
+
+        # 3. Get RMIA signals (placeholder)
+        rmia_signals = pd.read_csv(
+            "examples/ensemble_attack_example/data/attack_data/og_rmia_train_meta_pred.csv"
+        )  # Placeholder for RMIA features
+
+        continuous_features = df_input.loc[
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/blending.py:76]: Nope. Not sure why I did that.

File: src/midst_toolkit/attacks/ensemble/XGBoost.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+
+        Args:
+            x: Input features as a DataFrame.
+            y: Target variable as a numpy array.
+            use_gpu: Whether to use GPU acceleration. Defaults to False.
+        """
+        self.x = x
+        self.y = y
+        self.use_gpu = use_gpu
```
Comment (fatemetkl) [inline src/midst_toolkit/attacks/ensemble/XGBoost.py:38]: I really like this class attribute, but I only see it used in a commented line. Maybe it us used somewhere else that I am missing.

File: src/midst_toolkit/attacks/ensemble/blending.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Initializes the Blending++ attack with specified meta-classifier type and data configurations.
+        1. meta_classifier_type: Type of classifier to use ('lr' for Logistic Regression, 'xgb' for XGBoost).
+        2. data_configs: Configuration dictionary containing metadata about the dataset (e.g., column data type).
+        """
+        if meta_classifier_type not in ["lr", "xgb"]:
+            raise ValueError("meta_classifier_type must be 'lr' or 'xgb'")
+        self.meta_classifier_type = meta_classifier_type
+        self.data_configs = data_configs
+        self.meta_classifier_ = None  # The trained model, underscore denotes fitted attribute
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/blending.py:34]: Why would it?

File: src/midst_toolkit/attacks/ensemble/XGBoost.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            The best Scikit-learn Pipeline with optimized hyperparameters.
+
+        """
+
+        def objective(trial: Trial) -> float:
+            return self._evaluate_pipeline_cv(trial, num_kfolds=num_kfolds)
+
+        study = optuna.create_study(
+            direction="maximize",
+            sampler=optuna.samplers.TPESampler(n_startup_trials=10, seed=np.random.randint(1000)),
```
Comment (fatemetkl) [inline src/midst_toolkit/attacks/ensemble/XGBoost.py:135]: I see this is the random seed used in the original code, but as David suggested we can make it an option for the class and maybe pass here the user specified random_seed (it could be the one user sets in config.yaml).

File: src/midst_toolkit/attacks/ensemble/XGBoost.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                    "xgboost",
+                    xgb.XGBClassifier(
+                        n_estimators=100,
+                        eta=trial.suggest_float("eta", 0.0001, 0.1, log=True),
+                        max_depth=trial.suggest_int("max_depth", 3, 10),
+                        subsample=trial.suggest_float("subsample", 0.1, 1),
+                        colsample_bytree=trial.suggest_float("colsample_bylevel", 0.5, 1),
+                        reg_alpha=trial.suggest_categorical("reg_alpha", [0, 0.1, 0.5, 1, 5, 10]),
+                        reg_lambda=trial.suggest_categorical("reg_lambda", [0, 0.1, 0.5, 1, 5, 10, 100]),
+                        tree_method="auto",
```
Comment (fatemetkl) [inline src/midst_toolkit/attacks/ensemble/XGBoost.py:80]: Out of curiosity, was there an issue with `tree_method="auto" if not use_gpu else "gpu_hist",` ?

File: src/midst_toolkit/attacks/ensemble/distance_features.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def calculate_domias(df_input: pd.DataFrame, df_synth: pd.DataFrame, df_ref: pd.DataFrame) -> np.ndarray:
+    """
+    Compute DOMIAS density-ratio-based scores for test data.
+
+    Args:
+        df_input: Test data to evaluate (without labels).
+        df_synth: Synthetic data.
+        df_ref: Reference (real) population data.
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/distance_features.py]: Hold out. @fatemetkl could you please verify?

File: src/midst_toolkit/attacks/ensemble/blending.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lotif) [reply src/midst_toolkit/attacks/ensemble/blending.py:34]: The trailing "_" is not a standard naming convention, but I guess mypy does not have that one covered. If there is not a good reason for having it, I'd say we should remove it.

**Review by fatemetkl — COMMENTED** [review]
Really nice implementation of the blending++ pipeline! I also double-checked your implementation against the original ensemble codebase, and it matches perfectly 🙂.

I’ve left some comments, and I’ll address the ones where you’ve tagged me.
One suggestion (feel free to ignore): to clarify the different dataframes in this pipeline, in addition to docstrings, we could also reference the descriptions and diagrams in the example README.

File: examples/ensemble_attack_example/data_configs/info.json [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        5,
+        6,
+        7
+    ],
+    "cat_col_idx": [],
+    "target_col_idx": [
+        3
+    ],
+    "file_type": "csv",
+    "data_path": "/projects/aieng/midst_competition/data/tabddpm/tabddpm_1/train.csv",
```
[the remark this answers is not in the document]
Comment (fatemetkl) [reply examples/ensemble_attack_example/data_configs/info.json]: I think this comment no longer applies to this PR considering the recent updates. I have addressed this issue in my branch (upcoming PR), where we actually use this data_path.

File: examples/ensemble_attack_example/real_data_collection.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        data_processing_config: Configuration dictionary containing data specific information.
+
+    Returns:
+        pd.DataFrame: The specified dataset in this setting.
+    """
+    assert data_split in [
+        "train",
+        "dev",
+        "final",
+    ], "data_split should be one of 'train', 'dev', or 'final'."
```
[the remark this answers is not in the document]
Comment (fatemetkl) [reply examples/ensemble_attack_example/real_data_collection.py]: Nice suggestions, @lotif! Since these are my commits from the previous PR, I’ll address all of the Enum-related comments in my next PR, if that makes sense.

File: src/midst_toolkit/attacks/ensemble/distance_features.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    # Evaluate input points under both densities
+    p_ref = density_ref(input)
+    p_synth = density_synth(input)
+
+    # Density ratio. The higher the ratio, the more likely the point is synthetic (not in the real data)
+    ratio = np.divide(p_synth, p_ref, out=np.zeros_like(p_synth), where=p_ref > 0)
+
+    # Scale to [0, 1]
+    pred_proba_domias = MinMaxScaler().fit_transform(ratio.reshape(-1, 1)).ravel()
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/distance_features.py]: It's essentially undoing what `.reshape(-1, 1)` is doing, and returning a 1D array. Reshaping in the first place because `MinMaxScaler().fit_transform` takes in a 2D array.

File: src/midst_toolkit/attacks/ensemble/blending.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        gower_features = calculate_gower_features(
+            df_input=df_input, df_synthetic=df_synth, categorical_column_names=cat_cols
+        )
+
+        # 2. Get DOMIAS predictions
+        domias_features = calculate_domias_score(df_input=df_input, df_synthetic=df_synth, df_reference=df_ref)
+
+        # 3. Get RMIA signals (placeholder)
+        rmia_signals = pd.read_csv(
+            "examples/ensemble_attack/data/attack_data/og_rmia_train_meta_pred.csv"
```
[the remark this answers is not in the document]
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/blending.py:86]: I'd rather keep it like this for now because I'm not passing config.yaml to blending.py and don't plan on doing it. It will be resolved soon though!

File: src/midst_toolkit/attacks/ensemble/XGBoost.py (hunk 5 above)
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/XGBoost.py:39]: I should've uncommented that line before pushing the code. My bad.

File: src/midst_toolkit/attacks/ensemble/XGBoost.py (hunk 8 above)
Comment (sarakodeiri) [reply src/midst_toolkit/attacks/ensemble/XGBoost.py:82]: Nope. I wanted it to run on my local without taking up GPU space. Should've just changed `use_gpu` to false in config.yaml :)

File: mypy.ini [hunk 14]
```diff
@@ -36,3 +36,6 @@ ignore_missing_imports = True
 
 [mypy-category_encoders.*]
 ignore_missing_imports = True
+
+[mypy-gower.*]
+ignore_missing_imports = True
```
Comment (lotif) [inline mypy.ini:41]: Why do we need this?

File: examples/ensemble_attack_example/config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply examples/ensemble_attack_example/config.yaml]: I see. Perhaps also not something about this "placeholder" being temporary

File: src/midst_toolkit/attacks/ensemble/distance_features.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+from scipy.stats import gaussian_kde
+from sklearn.preprocessing import MinMaxScaler
+
+
+def calculate_gower_features(df_input: pd.DataFrame, df_synth: pd.DataFrame, cat_cols: list) -> pd.DataFrame:
+    """
+    Computes Gower distance-based features for a target dataframe against a synthetic one.
+
+    Args:
+        df_input: The dataframe to generate features for (e.g., meta classifier train set).
```
[the remark this answers is not in the document]
Comment (emersodb) [reply src/midst_toolkit/attacks/ensemble/distance_features.py]: The computation of this Gower distance has a very specific purpose as it is being done as part of constructing an attack rather than **just** computing a distance. Since each of the inputs probably has a specific meaning within the attack construction, documenting that is helpful for a user to give this function the correct dataframes for feature construction.

File: src/midst_toolkit/attacks/ensemble/distance_features.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply src/midst_toolkit/attacks/ensemble/distance_features.py]: This type of thing is important to document here, as it ensures the user doesn't make errors when applying the attack.

**Review by fatemetkl — APPROVED** [review]
Looks great to me! :rocket:
